File: python/Single Number III.py
class Solution(object):
    def singleNumber(self, nums):
        xor = 0
        for num in nums: xor ^= num
        xor = xor & (xor - 1) ^ xor
        a = b = 0
        for num in nums:
            if xor & num:
                a ^= num
            else:
                b ^= num
        return [a, b]
        

File: cpp/Single Number III.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum ^= nums[i];
        }
        int n = sum & (~(sum - 1)), a = 0, b =0;
        for (int i = 0; i < nums.size(); i++) {
            if (n & nums[i]) {
                a ^= nums[i];
            } else {
                b ^= nums[i];
            }
        }
        vector<int> result = {a, b};
        return result;
    }
};
